[MEMKEYDB]Feat add building memkind

michalbiesek opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pmem/redis/168)
<!-- Reviewable:end -->

Commit:
Add basic support to memkind allocator

- compile against libmemkind: make MALLOC=memkind

Co-authored-by: jschmieg <jakub.schmiegel@intel.com>

Commit:
Add memkind defrag implementation

Co-authored-by: jschmieg <jakub.schmiegel@intel.com>

kasiawasiuta reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 7 of 7 files at r1.
> *[Reviewable](https://reviewable.io/reviews/pmem/redis/168#-:-M4ny0E256uMHrXkjR5T:bg7igkv)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @jschmieg)



<!-- Sent from Reviewable.io -->

guoanwu commented on src/Makefile:
```diff
@@ -70,6 +70,10 @@ ifeq ($(USE_JEMALLOC),no)
 	MALLOC=libc
 endif
 
+ifeq ($(USE_MEMKIND),yes)
```
we can let the memkind as the default malloc and update the read me.

michalbiesek replied:
Done. Memkind is used as default allocator

Commit:
Use memkind allocator as default

jschmieg commented:
Is it possible to specify the version of memkind in INFO MEMORY response, like it is done with jemalloc? Is this exposed from building process?
mem_allocator:jemalloc-5.1.0
mem_allocator:memkind

michalbiesek reviewed (commented):
This would require changes in libmemkind. The information is exposed from libmemkind and must to be know during redis compilation process. Redis doesn't use the pkgconfig but I willtake a look for what should be change in libmemkind to achieve this. Just let me know it if it is required here - if yes then if should be added now or could be done in separate PR.

> *[Reviewable](https://reviewable.io/reviews/pmem/redis/168#-:-M4yCYIX24yJN5u0p7rG:b-qk434h)* status: 5 of 7 files reviewed, 1 unresolved discussion (waiting on @guoanwu, @jschmieg, and @kasiawasiuta)



<!-- Sent from Reviewable.io -->

Commit:
Add PMEM mechanism to zmalloc module

- use MEMKIND_DAX_KMEM for PMEM allocations
- use threshold to distinguish allocaitons between DRAM and PMEM

Co-authored-by: jschmieg <jakub.schmiegel@intel.com>

Commit:
Update INFO command

- provide information related to PMEM settings

Co-authored-by: jschmieg <jakub.schmiegel@intel.com>

Commit:
Add PMEM config test

Check correct setup of PMEM configuration
To run test, call:
- make CFLAGS=-DREDIS_TEST
- ./redis-server test pmem

michalbiesek commented:
Merged in memkeydb.
